### PR TITLE
Spectral guide

### PR DESCRIPTION
**Pull request by jerzyn**
Add a guide on how validate API specifications using spectral.

### CODE REVIEW

**Comment by jerzyn** [comment]
First draft. Treat it as a kick-off for the guide.

@zdne not sure how to add this inside the gitbook thingy. Anything in particular needed?

File: README.md [hunk 1]
```diff
@@ -55,7 +55,7 @@ npm install -g @stoplight/spectral
 yarn global add @stoplight/spectral
 ```
 
-Once installed, to verify your OAS file with spectral execute `spectral lint PATH_TO_YOUR_OAS -r spectral.yml`
+Once installed, to verify your OAS file with spectral execute `spectral lint PATH_TO_YOUR_OAS` having the `.spectral.yaml` file inside the directory from which you are calling the command.
```
Comment (zdne) [inline README.md]: `PATH_TO_YOUR_OAS` --> `<oas-file>` 

Also I doubt that people will have the current directory as the dir where the guidelines (`.spectral.yaml` ) is – maybe we should keep the `-r api-guidelines/.spectral` there for better illustration (= being more explicit)

File: general-guidelines/validating-api-descriptions.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# Validating API Description
+
+API description files (e.g. OpenAPI specification for REST APIs) SHOULD be validated to verify whether they comply with this guidelines. In order to do that you can utilize [Spectral](https://github.com/stoplightio/spectral) ruleset implementing them.
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: I'd rephrase it the other way around – we are not doing governance here – we are helping others to follow the guidelines.

Something like this: 

> If you want to follow adidas API design guidelines for REST APIs you MAY validate your OAS file using ....

File: general-guidelines/validating-api-descriptions.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+If the file you want to lint is on your computer, you'll need to mount the directory where the file resides as a volume
+
+```
+docker run --rm -it -v $(pwd):/tmp stoplight/spectral lint "/tmp/file.yaml"
+```
+
+## Using Spectral
+
+Once installed Spectral you can validate any `YAML` or `JSON` file according to a given set of rules. Spectral has a predefined set of rules validating OpenAPI 2.x (Swagger) and OpenAPI 3.x files.
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: Maybe change 

> Once installed Spectral you can validate any **`YAML` or `JSON` file according** ... 

to something like: 

> Once installed Spectral, you can validate **a OAS file (in YAML or JSON format)** ...

File: general-guidelines/validating-api-descriptions.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Spectral comes with a CLI and can be run from your command line:
+
+```
+spectral lint MY_SPEC_FILE.yaml
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: **Inconsistency**

`PATH_TO_YOUR_OAS` --> `<oas-file>`

`MY_SPEC_FILE.yaml` --> `<oas-file>`

File: general-guidelines/validating-api-descriptions.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+For more details about how to utilize Spectral CLI you can check the CLI built-in help:
+
+```
+spectral lint -h
+```
+
+or go to the official [Spectral documentation](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/cli.md).
+
+Spectral can also be used from within Javascript. For details on how to accomplish this, please refer to the [documentation](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/javascript.md).
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: `JavaScript`

File: general-guidelines/validating-api-descriptions.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+## Validating with Adidas API Guidelines
+
+To check whether your API Specification complies with Adidas API Guidelines copy the `.spectral` file from this repository ([here](https://github.com/adidas/api-guidelines/blob/master/.spectral.yml)).
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: I'd suggest against copying the file around. People should always use the latest from the repository. 

Eventually we could create a npm package from this repository so people would do something like `$ npm install -g adidas-api-guidelines@2.3.0` and then `$ api-guidelines  check <oas-file>` ...

File: general-guidelines/validating-api-descriptions.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+Once it is in the same directory as the one from which you are calling spectral it will be used automatically.
+
+``` 
+spectral lint my-api-spec.yaml
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: Consistency 

`my-api-spec.yaml` --> `<oas-file>`

File: general-guidelines/validating-api-descriptions.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+### Validation problems
+
+Spectral defines 4 levels of problem severity:
+
+1. error - the specification either does not pass standard YAML/JSON structure validation, standard OAS2/OAS3 validation or is not compliant with core adidas guidelines.
+2. warning - the specification lacks certain important validation check (e.g. `hosts` for OAS2) which are not required, but you should be cautious about them.
+3. information - there are some best practices which you did not implement in your specification and should consider doing so.
+4. hint - you should be aware that there are some additional best practices which you can consider implementing in your specification.
+
+Currently any problem of severity of error or warning will cause a failure status code of `1`. This means that any `error` or `warning` in your specification will prevent your CI/CD pipeline from succeeding. During design process you may want to ignore certain warnings, such as the runtime information (e.g. `hosts` or `servers`). In order to do that you can add a `--skip-rule` flag:
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: OK, how do I learn about the rules (ids) I can skip? From the error output? Also a link to spectral docs (if exists) could help!

File: README.md (hunk 1 above)
Comment (jerzyn) [reply README.md]: The path to your oas is fixed :) 

As for the other one - I'll think about packaging guidelines.

File: general-guidelines/validating-api-descriptions.md (hunk 6 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: you want the npm thing now or can we leave the first release as it is and the second iteration will already package it?

File: general-guidelines/validating-api-descriptions.md (hunk 8 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: makes sense to add the rules

File: general-guidelines/validating-api-descriptions.md (hunk 7 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: fixed

File: general-guidelines/validating-api-descriptions.md (hunk 5 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: fixed

File: general-guidelines/validating-api-descriptions.md (hunk 4 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: fixed

File: general-guidelines/validating-api-descriptions.md (hunk 3 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: changed

File: general-guidelines/validating-api-descriptions.md (hunk 2 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: changed

File: general-guidelines/validating-api-descriptions.md (hunk 8 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: I added adidas rules. Thinking whether it makes sense to add also OAS2/3 rules listing (I don't think it's listed on spectral documentation)

Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: found and added

**Comment by jerzyn** [comment]
FYI the PR also adds spectral people: @P0lip and @philsturgeon to the Acknowledgement section. Thanks guys for helping on this!

File: README.md (hunk 1 above)
Comment (jerzyn) [reply README.md]: done

File: general-guidelines/validating-api-descriptions.md (hunk 6 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: changed. ill package the guidelines in another PR

File: README.md [hunk 9]
```diff
@@ -55,7 +55,7 @@ npm install -g @stoplight/spectral
 yarn global add @stoplight/spectral
 ```
 
-Once installed, to verify your OAS file with spectral execute `spectral lint PATH_TO_YOUR_OAS` having the `.spectral.yaml` file inside the directory from which you are calling the command.
+Once installed, to verify your OAS file with spectral execute `spectral lint <oas-file> -r path/.spectral.yaml` where `path/.spectral.yaml` indicated the location `.spectral.yaml` file.
```
Comment (zdne) [inline README.md]: maybe: 

`path/.spectral.yaml`  -> `<adidas-api-guidelines-folder>/.spectral.yaml` ?

File: general-guidelines/validating-api-descriptions.md [hunk 10]
```diff
@@ -46,16 +46,14 @@ spectral lint -h
...
 ## Validating with Adidas API Guidelines
 
-To check whether your API Specification complies with Adidas API Guidelines copy the `.spectral` file from this repository ([here](https://github.com/adidas/api-guidelines/blob/master/.spectral.yml)).
-
-Once it is in the same directory as the one from which you are calling spectral it will be used automatically.
+To check whether your API Specification complies with Adidas API Guidelines you will need the `.spectral.yaml` file from this repository ([here](https://github.com/adidas/api-guidelines/blob/master/.spectral.yml)).
 
 ``` 
-spectral lint my-api-spec.yaml
+spectral lint <oas-file> -r path/.spectral.yaml
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: `path/.spectral.yaml`  -> `<adidas-api-guidelines-folder>/.spectral.yaml`

File: general-guidelines/validating-api-descriptions.md [hunk 11]
```diff
@@ -72,3 +70,36 @@ Currently any problem of severity of error or warning will cause a failure statu
...
+
+## Spectral rules list
+
+The documentation on Spectral general OAS and specific rules for OAS2 and OAS3 can be found here: https://github.com/stoplightio/spectral/blob/develop/docs/reference/openapi-rules.md. 
+
+Adidas specific rules are listed below:
+
+### Adidas general rules
+
+* `paths-camelCase` - All YAML/JSON paths MUST follow camelCase.
```
Comment (zdne) [inline general-guidelines/validating-api-descriptions.md]: Just an idea: Shouldn't we prefix all adidas rules with something like `adi-` or `adidas-` to distinguish from the others? (Does not have to be in this PR)

Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: thats something to discuss i think

File: README.md (hunk 9 above)
Comment (jerzyn) [reply README.md]: updated

File: general-guidelines/validating-api-descriptions.md (hunk 10 above)
Comment (jerzyn) [reply general-guidelines/validating-api-descriptions.md]: updated
